source/code/plugins/VMInsightsEngine.rb

```diff
@@ -0,0 +1,323 @@
+# frozen_string_literal: true
+
+module VMInsights
+
+    require_relative 'VMInsightsDataCollector.rb'
+
+    class MetricsEngine
+
+        require 'json'
+        require 'thread'
+
+        def initialize
+            @thread = nil
+        end
+
+        def start(config, &cb)
+            raise ArgumentError, 'config is nil' if config.nil?
+            raise ArgumentError, "config is not kind of #{Configuration}" unless config.kind_of? Configuration
+            raise ArgumentError, 'proc required' if cb.nil?
+            raise RuntimeError, 'already started' unless @thread.nil?
+            @thread = PollingThread.new config.poll_interval, config.computer, config.log, config.data_collector, cb
+            nil
+        end
+
+        def stop
+            return if @thread.nil?
+            @thread = nil if @thread.stop
+        end
+
+        def running?
+            (! @thread.nil?) && @thread.status
+        end
+
+        class Configuration
+            def initialize(computer, log, data_collector)
+                raise ArgumentError unless log
+                raise ArgumentError, "#{data_collector.class.name}" unless data_collector.kind_of? IDataCollector
+                @poll_interval = 60 # seconds
+                @computer = computer
+                @data_collector = data_collector
+                @log = log
+            end
+
+            def poll_interval=(v)
+                raise ArgumentError unless (v.kind_of? Numeric) && (v.real?) && (v >= 1)
+                @poll_interval = v
+            end
+
+            attr_reader :poll_interval, :computer, :data_collector, :log
+
+        end # class Configuration
+
+    private
+
+        class PollingThread < Thread
+            def initialize(interval, computer, log, data_collector, cb)
+                @mutex = Mutex.new
+                @condvar = ConditionVariable.new
+                @run = true
+                @log = log
+                @data_collector = data_collector
+                @cb = cb
+                @saved_exception = SavedException.new
+                @saved_cpu_exception = SavedException.new
+                @cummulative_data = CummulativeData.new
+                super() {
+                    @cummulative_data.initialize_from_baseline @data_collector.baseline
+                    MetricTuple.computer computer
+                    begin
+                        @log.info "Starting polling loop at #{interval} second interval"
+                        # try to keep as close to the polling interval, independent of
+                        # any delay waking up from the sleep, collecting data, or
+                        # emitting the message.
+                        expected_wakeup_time = Time.now
+                        while @run do
+                            expected_wakeup_time += interval
```

Comment:
This approach attempts to keep the schedule and will do just fine as long as the clock on the machine does not jump back/forward in time, as it relies on checks against Time.now every iteration.

An alternative implementation is to measure the elapsed time of data collection (using something that does not depend on Time.now but on tick differences, and then computing the sleep time. That way we can avoid the use of Time.now and be less susceptible to clock fluctuations.

That said, the data output is tagged with Time.now, so the current implementation should be good enough.

Reply:
Yes, "large" clock changes, both forward and backwards, mess up both the polling and rate calculations. As I recall, it is also an issue for the Windows version. When the issue was identified during design, we decided to defer addressing it.